### PR TITLE
[HttpFoundation] Add Psr6SessionHandler

### PR DESCRIPTION
**Pull request by atailouloute**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | https://github.com/orgs/symfony/projects/1#card-30499487
| License       | MIT
| Doc PR        | TODO

Add a PSR6 based session handler
By using this handler we can benefit from the cache `SodiumMarshaller` implemented in https://github.com/symfony/symfony/pull/35019, to encrypt/decrypt session data

Here is an example of how this can be used:

```yaml
# services.yaml
services:
    # ...
    Symfony\Component\HttpFoundation\Session\Storage\Handler\Psr6SessionHandler:
        arguments:
            $cache: '@cache.app'
```

```yaml
# packages/framework.yaml
framework:
    # ...
    session:
        handler_id: Symfony\Component\HttpFoundation\Session\Storage\Handler\Psr6SessionHandler
```

# Task list
- [ ] Add unit test
- [ ] Add doc

/cc @nicolas-grekas

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Thanks for working on this. Please compare your implementation to #23321. We'd need to ensure all failures of the cache adapter are properly turned to exceptions.

**Comment by atailouloute** [comment]
I compared both implementation, and technically they are nearly the same

I also noticed that the Psr6 handler is rejected twice as some cache adapter are by-design unreliable for session handling (lack of locking ...)

As the main benefit is to be able to encrypt session data, wdyt about experimenting also a `SodiumSessionHandler`?

**Comment by derrabus** [comment]
I think that remark here is still valid: https://github.com/symfony/symfony/pull/19193#issuecomment-228842802

> A cache can rightfully loose items. A session handler can't.

File: src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/Psr6SessionHandlerTest.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    const TTL = 100;
+    const PREFIX = 'pre';
+
+    /**
+     * @var Psr6SessionHandler
+     */
+    private $handler;
+
+    /**
+     * @var \PHPUnit_Framework_MockObject_MockObject|CacheItemPoolInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/Psr6SessionHandlerTest.php]: MockObject should use a namespace now (same below)

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/Psr6SessionHandler.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+     *
+     * @param CacheItemPoolInterface $cache   A Cache instance
+     * @param array                  $options An associative array of cache options
+     */
+    public function __construct(CacheItemPoolInterface $cache, array $options = [])
+    {
+        $this->cache = $cache;
+
+        if ($diff = array_diff(array_keys($options), ['prefix', 'ttl'])) {
+            throw new \InvalidArgumentException(sprintf('The following options are not supported "%s"', implode(', ', $diff)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/Psr6SessionHandler.php]: `sprintf('The following options are not supported by %s: "%s".', static::class, implode('", "', $diff))`

File: src/Symfony/Component/HttpFoundation/composer.json [hunk 3]
```diff
@@ -23,10 +23,12 @@
     "require-dev": {
         "predis/predis": "~1.0",
         "symfony/mime": "^4.4|^5.0",
-        "symfony/expression-language": "^4.4|^5.0"
+        "symfony/expression-language": "^4.4|^5.0",
+        "psr/cache": "~1.0"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/composer.json]: alpha order

File: src/Symfony/Component/HttpFoundation/composer.json [hunk 4]
```diff
@@ -23,10 +23,12 @@
...
     },
     "suggest" : {
-        "symfony/mime": "To use the file extension guesser"
+        "symfony/mime": "To use the file extension guesser",
+        "psr/cache": "To use the Psr6SessionHandler"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/composer.json]: let's remove this, we try to not use "suggest" much

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/Psr6SessionHandler.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        // not required here because cache will auto expire the records anyhow.
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function updateTimestamp($sessionId, $data)
+    {
+        $cacheItem = $this->cache->getItem($this->prefix.$sessionId);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/Psr6SessionHandler.php]: this one is quite unfortunate - we don't have a way to update a TTL without fetching the whole value...

**Review by nicolas-grekas — COMMENTED** [review]
> I think that remark here is still valid: https://github.com/symfony/symfony/pull/19193#issuecomment-228842802
> > A cache can rightfully loose items. A session handler can't.

yep, that's still a valid argument. We can decide to ignore it, but we'd need a reason.
Before trying to do so, would it make sense to use another approach?
The linked card says:
> Add SodiumSessionHandler to encrypt session data XOR add a PSR-6 handler

I'm now thinking about a 3rd way: adding a MarshallingSessionHandler that transforms the session data string before passing it to the decorated session handler:

`new MarshallingSessionHandler($someAbstractSessionHandler, $someMarshallerInterfaceInstance)`

That'd allow reusing the code of the marshallers in the Cache component, which is the real motivation for this card.

WDYT?

**Comment by atailouloute** [comment]
Closed in favor of #35804

**Pull request closed without merging**